File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[0][0]:
            return 0
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0] * n for _ in range(m)]
        
        for i in range(n):
            if not obstacleGrid[0][i]:
                dp[0][i] = 1
            else:
                break
        for i in range(m):
            if not obstacleGrid[i][0]:
                dp[i][0] = 1
            else:
                break
        
        for i in range(1, m):
            for j in range(1, n):
                if obstacleGrid[i][j]:
                    continue
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
        
        return dp[m - 1][n - 1]
        
        
        

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
private:
    int dp(int i, int j, vector<vector<int>>& obstacleGrid, vector<vector<int>>& memo){
        if (i >= 0 && j >= 0 && obstacleGrid[i][j]) return 0;
        if (i == 0 && j == 0) return 1;
        if (i < 0 || j < 0) return 0;
        if (memo[i][j] != -1) return memo[i][j];
        return memo[i][j] = dp(i - 1, j, obstacleGrid, memo) + dp(i, j - 1, obstacleGrid, memo);
    }
    
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        vector<vector<int>> memo(m, vector<int>(n, -1));
        
        return dp(m - 1, n - 1, obstacleGrid, memo);
    }
};